---
# 단어 빈도를 이용한 벡터화 (1) Bag of Words
---

머신러닝 모델은 기계이기 때문에, 텍스트보다는 수치화된 숫자를 더 잘 처리할 수 있습니다.  
이런 이유로 기계가 자연어 처리를 원활히 할 수 있도록, 자연어 처리에서는 전처리 과정에서 텍스트를 숫자 벡터로 변환하는 벡터화(Vectorization) 라는 과정을 거치게 됩니다.  

벡터화 방법으로는 크게 (1) 통계와 머신 러닝을 활용한 방법, 그리고 (2) 인공 신경망을 활용하는 방법 두 가지로 나눠볼 수 있는데요.  
인공 신경망을 활용하는 방법은 다음에 배워보도록 하고, 오늘은 전자의 방법을 통해 벡터화하는 방법에 대해서 배워보겠습니다.

---
### Bag of Words
---

Bag of Words는 자연어 처리(Natural Language Processing)나 정보 검색(Information Retrieval)에서 쓰이는 매우 간단한 단어 표현 방법으로,  
### **문서 내의 단어들의 분포를 보고 이 문서의 특성을 파악하는 기법입니다.** 
줄여서 간단히 BoW라고 표기하기도 합니다.

아직은 설명이 아리송하지요? 조금 더 쉽게 설명해볼게요.  
Bag of Words는 직역하면 "단어들의 가방"입니다.  
Bag of Words는 문서를 단어들의 가방으로 가정합니다.

<img src='https://d3s0tskafalll9.cloudfront.net/media/images/GN-2-L-1.max-800x600.png'>

어떤 한 문서에 등장하는 텍스트를 전부 단어 단위로 토큰화(tokenization)를 합니다.  
그러면 이제 수많은 단어들이 나올 텐데, 이 단어들을 전부 가방에 집어넣습니다.  
그리고 이 가방을 마구 흔들어요. 그렇게 되면 단어들의 순서는 무시되고 전부 섞이게 되겠지요?

그리고 문서에 사과라는 단어가 6번 등장했다면, 이 가방에는 사과라는 단어가 6개 있을 것입니다. Bag of Words는 중복을 제거하지 않고, 단어들을 카운트해요.  
### **결론적으로 Bag of Words는 단어들의 순서는 무시하고, 단어들의 빈도 정보는 그대로 보존해요.**

예를 들어볼게요! 다음과 같은 doc1이 있다고 해볼게요.

    
    doc1 = 'John likes to watch movies. Mary likes movies too.'
이를 BoW로 표현하면 아래와 같습니다.

    BoW1 = {"John":1, "likes":2, "to":1, "watch":1, "movies":2, "Mary":1, "too":1}
아래와 같이 doc2가 있다고 해보겠습니다.

    doc2 = 'Mary also likes to watch football games.'
이를 BoW로 표현하면 아래와 같겠군요!

    BoW2 = {"Mary":1, "also":1, "likes":1, "to":1, "watch":1, "football":1, "games":1}
위의 BoW에서 각 key는 단어, 각 value는 각 단어가 주어진 문서에 등장한 횟수에요. 이때, 문서 내 단어의 순서는 중요하지 않습니다. 그렇기 때문에 아래의 BoW는 BoW1과 본질적으로 동일합니다.

    BoW = {"too":1, "Mary":1, "movies":2, "John":1, "watch":1, "likes":2, "to":1}
    BoW1 = {"John":1, "likes":2, "to":1, "watch":1, "movies":2, "Mary":1, "too":1}

BoW 방식으로 문장을 벡터화한다면 I ate lunch. 와 Lunch ate I. 는 같은 문장이 됩니다.  
이렇게 어순에 따라 달라지는 의미를 반영하지 못한다는 점에서 BoW는 한계를 가집니다.

이번에는 두 문서를 합쳐서 새로운 문서를 정의하고 BoW를 구해볼게요!

    doc3 = 'John likes to watch movies. Mary likes movies too. Mary also likes to watch football games.'
이로부터 만들어진 BoW3은 아래와 같습니다.

    BoW3 = {"John":1, "likes":3, "to":2, "watch":2, "movies":2, "Mary":2, "too":1, "also":1, "football":1, "games":1};

---
# 단어 빈도를 이용한 벡터화 (2) Bag of Words 구현해보기
---

---
### keras Tokenizer 활용
---

케라스 토크나이저를 통해 Bag of Words를 만들어 보겠습니다.

In [1]:
from tensorflow.keras.preprocessing.text import Tokenizer

sentence = ["John likes to watch movies. Mary likes movies too! Mary also likes to watch football games."]

tokenizer = Tokenizer()
tokenizer.fit_on_texts(sentence) # 단어장 생성
bow = dict(tokenizer.word_counts) # 각 단어와 각 단어의 빈도를 bow에 저장

print("Bag of Words :", bow) # bow 출력
print('단어장(Vocabulary)의 크기 :', len(tokenizer.word_counts)) # 중복을 제거한 단어들의 개수

Bag of Words : {'john': 1, 'likes': 3, 'to': 2, 'watch': 2, 'movies': 2, 'mary': 2, 'too': 1, 'also': 1, 'football': 1, 'games': 1}
단어장(Vocabulary)의 크기 : 10


여기서 단어장(Vocabulary) 이라는 또 다른 개념을 이해해 보겠습니다.  
문헌에 따라 사전 또는 단어 집합이라고 불리는데요. 단어장이란 중복을 제거한 단어들의 집합을 말합니다.  
이는 Bag of Words랑은 다른 개념입니다. 위의 예제에서는 단어장의 크기가 10개네요!

---
### scikit-learn CountVectorizer 활용
---

In [2]:
from sklearn.feature_extraction.text import CountVectorizer

sentence = ["John likes to watch movies. Mary likes movies too! Mary also likes to watch football games."]

vector = CountVectorizer()
bow = vector.fit_transform(sentence).toarray()

print('Bag of Words : ', bow) # 코퍼스로부터 각 단어의 빈도수를 기록한다.
print('각 단어의 인덱스 :', vector.vocabulary_) # 각 단어의 인덱스가 어떻게 부여되었는지를 보여준다.

Bag of Words :  [[1 1 1 1 3 2 2 2 1 2]]
각 단어의 인덱스 : {'john': 3, 'likes': 4, 'to': 7, 'watch': 9, 'movies': 6, 'mary': 5, 'too': 8, 'also': 0, 'football': 1, 'games': 2}


CounterVectorizer의 vector.fit_transform(입력 문장).toarray()으로 출력되는 Bag of Words 결과를 보면  
각 단어의 빈도만 출력될 뿐, 어떤 단어의 빈도인지는 나오지 않습니다.  
[[1 1 1 1 3 2 2 2 1 2]]이라는 값만으로는 각 숫자가 어떤 단어의 빈도인지 알 수가 없지요?

이는 vector.vocabulary_를 통해서 각 단어에 부여된 인덱스를 확인하면 되는데요.  
주의할 점은 이 인덱스는 0부터 시작한다는 점입니다. 가령, also의 인덱스는 0이네요.  
그런데 BoW에서 0번째 위치한 값은 1이니까, also의 빈도는 1입니다. 또는 likes의 인덱스는 4인데,  
BoW에서 4번째 위치한 값은 3이니까 likes의 빈도는 3임을 알 수 있지요!

그리고 이 Bag of Words의 단어장의 크기는 아래와 같이 구하면 되겠지요!

In [3]:
print('단어장(Vocabulary)의 크기 :', len(vector.vocabulary_))

단어장(Vocabulary)의 크기 : 10


단어장이라는 개념은 Bag of Words를 위해 사용되는 개념이라기보다는, 자연어 처리에서 보편적으로 쓰이는 용어입니다.  
그런데 이를 계속 강조하는 이유는 지금부터 사용할 DTM(Document-Term Matrix) 이라는 개념 이해를 돕기 위해서입니다.  
지금 Bag of Words를 배우면서 이런 생각이 들지 않으셨나요? 그래서 이 개념을 도대체 어디서 사용할 수 있다는 거지? 이를 이해하기 위해서 DTM으로 가보겠습니다!

---
# 단어 빈도를 이용한 벡터화 (3) DTM과 코사인 유사도
---

DTM(Document-Term Matrix) 은 직역하면 문서-단어 행렬입니다.  
DTM은 여러 문서의 Bag of Words를 하나의 행렬로 구현한 것인데요.  
다시 말해 DTM은 각 문서에 등장한 단어의 빈도수를 하나의 행렬로 통합시킵니다.  
DTM은 문서를 행으로, 단어를 열로 가지는 행렬이지만,  
문헌에 따라서는 열을 문서로 하고 단어를 행으로 하여 TDM(Term-Document Matrix) 이라고 부르기도 합니다.

예를 들어, 아래와 같은 3개의 문서가 있다고 해보겠습니다.

>Doc 1: Intelligent applications creates intelligent business processes  
>Doc 2: Bots are intelligent applications  
>Doc 3: I do business intelligence  

이 3개의 문서로부터 얻을 수 있는 DTM은 아래와 같습니다.

<img src='https://d3s0tskafalll9.cloudfront.net/media/images/GN-2-L-2.max-800x600.png'>

위의 DTM을 보면 각 행은 각 문서를 나타내고 있습니다.  
그리고 각 열은 문서 3개의 통합 단어장(Vocabuary)에 있는 단어들로 구성되어져 있습니다.  
이렇게 구성하면 앞서 문서 1개로 BoW를 만들었을 때와는 달리, 각 행은 많은 값이 0으로 구성이 됩니다.

이때 각 행을 문서 벡터(document vector), 열을 단어 벡터(word vector)라고 부를 수 있는데요.   
문서의 수가 많아지면 많아질수록, 통합 단어장의 크기도 커지게 되어서 DTM은 결국 문서 벡터와 단어 벡터 모두 대부분의 값이 0이 되는 성질을 가지고 있습니다.

여러 문서들로 DTM을 만들었다면, 이제 Bag of Words로 무엇을 할 수 있는지를 이해할 차례입니다.  
DTM을 사용하면 각 문서들을 비교하고, 각 문서 간 유사도를 구할 수 있게 됩니다. 예를 들어볼까요?

> 문서1 : I like dog  
> 문서2 : I like cat  
> 문서3 : I like cat I like cat

위와 같이 3개의 문서가 있다고 해볼게요. 각 단어에 대한 인덱스를 cat은 0, dog는 1, I는 2, like는 3이라고 했을 때 생성되는 DTM의 형태는 아래와 같습니다.

<img src='https://d3s0tskafalll9.cloudfront.net/media/original_images/GN-2-L-3.png'>

이제 각 문서 벡터의 유사도를 구해보겠습니다.  
저자는 가장 보편적으로 사용되는 유사도 계산 방법인 코사인 유사도를 택했습니다.  
아래와 같이 각 문서 벡터와 코사인 유사도 함수를 정의합니다.



In [4]:
import numpy as np
from numpy import dot
from numpy.linalg import norm

doc1 = np.array([0,1,1,1]) # 문서1 벡터
doc2 = np.array([1,0,1,1]) # 문서2 벡터
doc3 = np.array([2,0,2,2]) # 문서3 벡터

def cos_sim(A, B):
    return dot(A, B)/(norm(A)*norm(B))

In [5]:
print(cos_sim(doc1, doc2)) #문서1과 문서2의 코사인 유사도
print(cos_sim(doc1, doc3)) #문서1과 문서3의 코사인 유사도
print(cos_sim(doc2, doc3)) #문서2과 문서3의 코사인 유사도

0.6666666666666667
0.6666666666666667
1.0000000000000002


<img src='https://wikidocs.net/images/page/24603/코사인유사도.PNG'>

$$
similarity=cos(Θ)=\frac{A⋅B}{||A||\ ||B||}=\frac{\sum_{i=1}^{n}{A_{i}×B_{i}}}{\sqrt{\sum_{i=1}^{n}(A_{i})^2}×\sqrt{\sum_{i=1}^{n}(B_{i})^2}}
$$

눈여겨볼만한 점은 **문서1과 문서2의 코사인 유사도**와 **문서1과 문서3의 코사인 유사도**가 같다는 점과 **문서2와 문서3의 코사인 유사도가 1**이 나온다는 것입니다.  
앞서 1은 두 벡터의 방향이 완전히 동일한 경우에 1이 나오며,  
코사인 유사도 관점에서는 유사도의 값이 최대임을 의미한다고 언급한 바 있습니다.

문서3은 문서2에서 단지 모든 단어의 빈도수가 1씩 증가했을 뿐입니다.  
다시 말해 한 문서 내의 모든 단어의 빈도수가 동일하게 증가하는 경우에는 기존의 문서와 코사인 유사도의 값이 1이라는 것입니다.  
<br>
이것이 시사하는 점은 무엇일까요? 예를 들어보겠습니다. 문서 A와 B가 동일한 주제의 문서. 문서 C는 다른 주제의 문서라고 해봅시다. 
<br>
그리고 문서 A와 문서 C의 문서의 길이는 거의 차이가 나지 않지만, 문서 B의 경우 문서 A의 길이보다 두 배의 길이를 가진다고 가정하겠습니다. 
<br> 
이런 경우 유클리드 거리로 유사도를 연산하면 문서 A가 문서 B보다 문서 C와 유사도가 더 높게 나오는 상황이 발생할 수 있습니다.  
이는 유사도 연산에 문서의 길이가 영향을 받았기 때문인데, 이런 경우 코사인 유사도가 해결책이 될 수 있습니다.  
<br>
코사인 유사도는 유사도를 구할 때 벡터의 방향(패턴)에 초점을 두므로 코사인 유사도는 문서의 길이가 다른 상황에서 비교적 공정한 비교를 할 수 있도록 도와줍니다.

---
# 단어 빈도를 이용한 벡터화 (4) DTM의 구현과 한계점
---

---
### scikit-learn CountVectorizer 활용
---

사이킷런의 CountVectorizer를 이용하여 DTM을 만드는 방법은 앞서 Bag of Words를 만드는 방법과 동일합니다.  
단지, 다수의 문서를 입력값으로 주면 됩니다. 다음은 3개의 문서에 대해서 DTM을 만드는 과정을 보여줍니다.

In [7]:
from sklearn.feature_extraction.text import CountVectorizer

corpus = [
    'John likes to watch movies',
    'Mary likes movies too',
    'Mary also likes to watch football games',    
]
vector = CountVectorizer()

print(vector.fit_transform(corpus).toarray()) # 코퍼스로부터 각 단어의 빈도수를 기록.
print(vector.vocabulary_) # 각 단어의 인덱스가 어떻게 부여되었는지를 보여준다.

[[0 0 0 1 1 0 1 1 0 1]
 [0 0 0 0 1 1 1 0 1 0]
 [1 1 1 0 1 1 0 1 0 1]]
{'john': 3, 'likes': 4, 'to': 7, 'watch': 9, 'movies': 6, 'mary': 5, 'too': 8, 'also': 0, 'football': 1, 'games': 2}


---
### DTM의 한계점
---

이쯤에서 DTM의 성질을 정리해보도록 하겠습니다.

DTM은 bag of words 기반으로 문서를 비교할 수 있는 행렬입니다.

그런데 DTM은 두 가지 한계점을 지닙니다.

>먼저, DTM에서 문서의 수와 단어의 수가 계속 늘어날수록, 행과 열은 대부분의 값이 0을 가진다는 특징이 있습니다. 이는 저장 공간 측면에서 낭비입니다. (워드 임베딩 챕터에서 배우겠지만, 지나친 차원의 크기는 차원의 저주라는 또 다른 문제를 발생시킵니다.)

>두 번째 한계점은 단어의 빈도에만 집중하는 방법 자체의 한계입니다.  
예를 들어, 영어 데이터를 가지고 DTM을 만들었을 때,  
불용어 'the'는 어떤 문서에서도 자주 등장하는 경향이 있습니다.  
그런데 문서1과 문서2에서 둘 다 'the'가 많이 등장했다고 하여 이 두 문서가 유사한 문서라고 봐도 될까요?

>사실 여러 문서에는 중요한 단어와 중요하지 않은 단어들이 혼재되어 있는데,  
'the'와 같이 아무 문서에나 자주 등장하는 단어들은 문서끼리 비교할 때 그다지 도움이 되지 않습니다.   
그렇다면 중요한 단어와 중요하지 않은 단어에 가중치를 따로 선별해서 주는 방법은 없을까요?

---
# 단어 빈도를 이용한 벡터화 (5) TF-IDF
---

---
### TF-IDF란
---

각 단어의 중요도를 판단하여 가중치를 주는 방법인 TF-IDF에 대해서 알아봅시다.  
TF-IDF는 모든 문서에서 자주 등장하는 단어는 중요도가 낮다고 판단하며, 특정 문서에서만 자주 등장하는 단어는 중요도가 높다고 판단하는 것입니다.

TF-IDF는 Term Frequency-Inverse Document Frequency의 약자입니다.  
한국어로 해석하면 '단어 빈도-역문서 빈도' 입니다. 추측하건대, '단어의 빈도' 와 '문서의 빈도의 역수' 를 활용하는 것 같습니다.  
  
TF-IDF는 불용어처럼 중요도가 낮으면서 모든 문서에 등장하는 단어들이 노이즈가 되는 것을 완화해 줍니다.  
주의할 점은 TF-IDF를 사용하는 것이 DTM을 사용하는 것보다 성능이 항상 뛰어나지는 않다는 점입니다.  
TF-IDF를 사용하기 위해서는 우선 DTM을 만든 뒤에 TF-IDF 가중치를 DTM에 적용합니다.

단어의 빈도를 의미하는 TF는 사실 이미 배웠습니다.  
DTM이 이미 TF 행렬이기 때문이죠! 그리고 DTM의 각 단어에 IDF 값을 곱해주면 TF-IDF 행렬이 완성됩니다.

---
### TF-IDF 계산하기
---

<img src='https://d3s0tskafalll9.cloudfront.net/media/original_images/GN-2-L-4.png'>  

위의 이미지는 y는 문서, x는 단어라고 할 때 TF-IDF의 수식을 보여줍니다.  
TF는 앞서 설명했듯이 각 문서에 등장하는 단어의 빈도를 의미하므로 DTM을 만들면 자연스럽게 해결됩니다.

위의 수식에서 TF 뒤에 곱해지는 log항이 IDF입니다.  
TF를 구하는 방법은 알겠는데, 역문서 빈도인 IDF는 어떻게 구할까요?  
아래의 문제를 통해 이해해봅시다.  

>Question.  
전체 문서의 수가 5개라고 해봅시다. 그리고 단어 'like'가 문서2에서 200번, 문서 3에서 300번 등장했다고 해봅시다.  
다른 문서에서 단어 'like'는 등장하지 않았습니다. 이때, 단어 'like'의 IDF는 몇일까요?

IDF를 구하기 위해서는 우선 문서 빈도 DF와 전체 문서의 수 N을 이해해야 합니다.  
단어 'like'의 DF를 계산해 봅시다. DF는 문서의 빈도이므로 단어 'like'의 DF는 2입니다.  
단어 'like'가 어떤 문서에서 몇 번 등장했는지는 DF를 구할 때는 중요하지 않습니다.  
단어 'like'가 몇 개의 문서에서 등장했는지가 중요하죠. 즉 단어 'like'가 문서2와 문서3에서 등장했기 때문에 단어 'like'의 DF는 2입니다.  
N은 전체 문서의 수입니다. 위에서 전체 문서의 수는 5개라고 했었지요? 결과적으로 단어 'like'의 IDF는
log5/2가 되겠습니다.  
<br>
보통 TF-IDF를 지원하는 파이썬 패키지들에서 log는 자연로그를 사용합니다.  
여기서도 자연로그를 사용한다고 가정합시다. 자연로그는 로그의 밑으로 자연상수 e(e=2.718281...) 를 사용하는 로그를 말합니다. 자연로그는 보통 ln으로 표현하는데요.  
  
결국 단어 'like'의 IDF 계산식은 ln5/2가 되겠네요. 이 값은 계산해보면 0.91629073187입니다. 문제를 풀었네요!

위의 문제에서 한 발짝 더 나아가볼까요?

>Question.  
문서2과 문서3에서의 단어 'like'의 TF-IDF값은 몇 일까요?

문서2에서의 TF는 200, 문서3에서의 TF는 300이었습니다.  
문서2의 단어 'like'의 TF-IDF를 구하면 200×ln5/2 , 즉 183.258146375입니다.  

문서3에서의 단어 'like'의 TF-IDF는 300×ln5/2 이고, 계산하면 274.887219562입니다.

문서3에서의 TF-IDF 값이 더 높네요.  
TF-IDF의 목적이 특정 문서에서만 자주 등장하는 단어는 중요도는 높다고 판단한다는 것이었으므로 어찌 보면 당연한 결과군요!

[유튜브 링크](https://www.youtube.com/watch?v=Rd3OnBPDRbM)

---
# 단어 빈도를 이용한 벡터화 (6) TF-IDF 구현하기
---

---
### TF-IDF 구현하기
---

In [8]:
from math import log
import pandas as pd

여기서는 3개의 문서를 사용할게요!

In [9]:
docs = [
  'John likes to watch movies and Mary likes movies too',
  'James likes to watch TV',
  'Mary also likes to watch football games',  
]

우선 DTM의 열을 만들기 위해서 문서 3개의 단어들이 모두 들어간 통합 단어장을 만들겠습니다.

In [10]:
vocab = list(set(w for doc in docs for w in doc.split()))
vocab.sort()
print('단어장의 크기 :', len(vocab))
print(vocab)

단어장의 크기 : 13
['James', 'John', 'Mary', 'TV', 'also', 'and', 'football', 'games', 'likes', 'movies', 'to', 'too', 'watch']


현재 단어장의 크기는 13이네요. 중복을 제거한 단어들의 개수가 13개라는 의미입니다.  
이제 총 문서의 수를 변수 N에다가 저장하겠습니다.  
우리는 이미 이 값이 3이라는 사실을 알고 있지요?

In [11]:
N = len(docs) # 총 문서의 수
N

3

이제 TF 함수, IDF 함수, TF-IDF 함수를 만들겠습니다.

여기서 IDF를 실제 구현할 때는 앞서 배운 식과는 다소 다른 식을 사용하게 됩니다. 실제로 많은 파이썬 패키지들이 앞에서 배운 식에서 조금씩 조정된, 서로 다른 식을 사용합니다.

우선, 여기서는 log 항의 분모에 1을 더해주겠습니다.  
이는 (Bag of Words를 사용할 때 일반적인 상황은 아니지만) 특정 단어가 전체 문서에서 등장하지 않을 경우에 분모가 0이 되는 상황을 방지하기 위함입니다.  

그리고 log 항에 1을 더해주겠습니다. 이는 log 항의 분자와 분모값이 동일해졌을 때,   
log의 진수가 1이 됨에 따라서 IDF의 값이 0이 되는 것을 방지하기 위함입니다.

In [12]:
def tf(t, d):
    return d.count(t)
 
def idf(t):
    df = 0
    for doc in docs:
        df += t in doc    
    return log(N/(df + 1)) + 1
 
def tfidf(t, d):
    return tf(t,d)* idf(t)

우선 TF 함수를 사용하여 DTM을 만들어봅시다.

In [13]:
result = []
for i in range(N): # 각 문서에 대해서 아래 명령을 수행
    result.append([])
    d = docs[i]
    for j in range(len(vocab)):
        t = vocab[j]
        
        result[-1].append(tf(t, d))
        
tf_ = pd.DataFrame(result, columns = vocab)
tf_

James  John  Mary  TV  also  and  football  games  likes  movies  to  too  \
0      0     1     1   0     0    1         0      0      2       2   2    1   
1      1     0     0   1     0    0         0      0      1       0   1    0   
2      0     0     1   0     1    0         1      1      1       0   1    0   

   watch  
0      1  
1      1  
2      1

각 단어의 IDF를 구해봅시다.

In [14]:
result = []
for j in range(len(vocab)):
    t = vocab[j]
    result.append(idf(t))

idf_ = pd.DataFrame(result, index = vocab, columns=["IDF"])
idf_

IDF
James     1.405465
John      1.405465
Mary      1.000000
TV        1.405465
also      1.405465
and       1.405465
football  1.405465
games     1.405465
likes     0.712318
movies    1.405465
to        0.712318
too       1.405465
watch     0.712318

IDF는 'likes'나 'to' 그리고 'watch'와 같이 모든 문서에 등장한 단어가 가장 낮은 값을 가집니다.  
그리고 그다음으로 두 개의 문서에 등장한 'Mary'가 그다음으로 낮은 값을 가지며,  
그 외의 1개의 문서에만 등장한 단어들은 가장 높은 값을 가지는 것을 볼 수 있습니다.  
이제 TF-IDF 행렬을 출력해봅시다. DTM에 있는 각 단어의 TF에 각 단어의 IDF를 곱해준 값입니다.

In [15]:
result = []
for i in range(N):
    result.append([])
    d = docs[i]
    for j in range(len(vocab)):
        t = vocab[j]
        
        result[-1].append(tfidf(t,d))

tfidf_ = pd.DataFrame(result, columns = vocab)
tfidf_

James      John  Mary        TV      also       and  football     games  \
0  0.000000  1.405465   1.0  0.000000  0.000000  1.405465  0.000000  0.000000   
1  1.405465  0.000000   0.0  1.405465  0.000000  0.000000  0.000000  0.000000   
2  0.000000  0.000000   1.0  0.000000  1.405465  0.000000  1.405465  1.405465   

      likes   movies        to       too     watch  
0  1.424636  2.81093  1.424636  1.405465  0.712318  
1  0.712318  0.00000  0.712318  0.000000  0.712318  
2  0.712318  0.00000  0.712318  0.000000  0.712318

---
### scikit-learn TFidVectorizer 활용
---

사이킷런으로 DTM을 만들기 위해서는 CountVectorizer를 사용했었습니다.  
이와 유사하게 사이킷런에서는 TF-IDF를 자동으로 계산하여 출력하는 TfidfVectorizer를 제공합니다.
  
향후 실습 중에 결괏값이 예상과 달라 혼란이 일어나는 일이 없도록 미리 언급하자면,  
사이킷런의 TfidfVectorizer 또한 이번에 배운 기본식에서 조정된 식을 사용합니다.  
간단히 언급하면, TfidfVectorizer는 위에서 우리가 파이썬으로 구현한 식에서 더 나아가 log 항의 분자에도 1을 더해주며,  
TF-IDF의 결과에 L2 Norm까지 추가로 수행한다는 점이 다릅니다.
  
사이킷런의 TfidfVectorizer를 통해 TF-IDF 행렬을 출력해봅시다.

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
  'John likes to watch movies and Mary likes movies too',
  'James likes to watch TV',
  'Mary also likes to watch football games',  
]

tfidfv = TfidfVectorizer().fit(corpus)
vocab = list(tfidfv.vocabulary_.keys()) # 단어장을 리스트로 저장
vocab.sort() # 단어장을 알파벳 순으로 정렬

# TF-IDF 행렬에 단어장을 데이터프레임의 열로 지정하여 데이터프레임 생성
tfidf_ = pd.DataFrame(tfidfv.transform(corpus).toarray(), columns = vocab)
tfidf_

also       and  football     games     james      john     likes  \
0  0.000000  0.321556  0.000000  0.000000  0.000000  0.321556  0.379832   
1  0.000000  0.000000  0.000000  0.000000  0.572929  0.000000  0.338381   
2  0.464997  0.000000  0.464997  0.464997  0.000000  0.000000  0.274634   

       mary    movies        to       too        tv     watch  
0  0.244551  0.643111  0.189916  0.321556  0.000000  0.189916  
1  0.000000  0.000000  0.338381  0.000000  0.572929  0.338381  
2  0.353642  0.000000  0.274634  0.000000  0.000000  0.274634

---
# LSA와 LDA (1) LSA
---

DTM과 TF-IDF 행렬같이 Bag of Words를 기반으로 한 표현 방법은 근본적으로 단어의 의미를 벡터로 표현하지 못한다는 한계를 가지고 있습니다.  
가령, 일반적인 검색 작업에서 특정 단어들이 포함된 문서들을 단순히 찾아내는 것은 아주 빠르게 해낼 수 있습니다.  
그런데 특정 단어(단어의 의미)와 관련된 문서 집합(문서의 주제)을 찾아내고 싶다면 어떨까요?  
단어들의 빈도가 아닌 의미와 주제를 알고 싶다면 어떻게 해야 할까요?  
이는 DTM이나 TF-IDF 만으로는 풀기 어려운 문제입니다. 이를 해결할 수 있는 LSA(Latent Semantic Analysis) 에 대해 알아봅시다.  

---
### LSA(Latent Semantic Analysis)
---

한국어로 잠재 의미 분석이라고 불리는 LSA(Latent Semantic Analysis) 는 전체 코퍼스에서 문서 속 단어들 사이의 관계를 찾아내는 자연어 처리 정보 검색 기술입니다.  

LSA를 사용하면 단어와 단어 사이, 문서와 문서 사이, 단어와 문서 사이의 의미적 유사성 점수를 찾아낼 수 있습니다.  
그리고 이 방법은 어떤 문서에서 특정 단어들의 빈도가 몇인지를 판단하는 것보다 효과적인 경우가 많습니다.

---
### 특잇값 분해
---

LSA를 이해하려면 먼저 선형대수학의 특잇값 분해(Singular Value Decompotion) 에 대해 이해할 필요가 있습니다.  
우선 행렬에 대한 용어를 파악해야 합니다.  

<img src= 'https://d3s0tskafalll9.cloudfront.net/media/original_images/GN-2-L-9.png'>

특잇값 분해(Singular Value Decomposition, SVD) 란  
m × n 크기의 임의의 사각 행렬 A를 위의 그림의 Full SVD와 같이  
특이 벡터(singular vector)의 행렬와 특잇값(singular value)의 대각행렬로 분해하는 것을 말합니다. 

특잇값 분해 중에서 조금 특별한 특잇값 분해가 있습니다.  
특잇값 가운데 가장 큰(다시 말해, 가장 중요한) t개만 남기고 해당 특잇값에 대응되는 특이 벡터(singular vector) 들로  
행렬 A를 근사(approximate) 하도록 하면, 이를 절단된 특잇값 분해(Truncated SVD) 라고 합니다.

Truncated SVD를 수행하면 행렬 Σ의 대각 원솟값 중에서 상윗값 t개만 남게 되며, U행렬과 V행렬의 t열까지만 남습니다.  
이로 인해 세 행렬에서 값(정보)의 손실이 일어나 기존의 행렬 A를 정확히 복구할 수는 없게 됩니다.

여기서 t는 하이퍼파라미터입니다. 하이퍼파라미터란 사용자가 직접 값을 선택하며 성능에 영향을 주는 매개변수를 말합니다.  
t를 크게 잡으면 기존의 행렬 A로부터 다양한 의미를 가져갈 수 있지만, 노이즈를 제거하려면 t를 작게 잡아야 합니다.

------
## - 고유벡터  
정방 행렬 A를 선형 변환으로 봤을 때, 선형 변환 A에 의한 변환 결과가 자기 자신의 상수 배가 되는 0이 아닌 벡터를 고유벡터(eigenvector)라고 하고,  
이 상수배 값을 고유값(eigenvalue)이라고 합니다. 고유값, 고유 벡터는 정방 행렬에 대해서만 정의됩니다.  
다시 말해, 정방 행렬 A에 대해서 Av = λv를 만족하는 0이 아닌 열벡터 v를 고유 벡터, 상수 λ를 고유값이라고 합니다.

## - 전치행렬  
원래의 행렬에서 행과 열을 서로 맞바꾼 행렬  
  <img src='https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FbMlRFI%2FbtqBUUqRZI2%2FCsRg3y4xI9iL8mB6QxfDlk%2Fimg.jpg'>

## - 단위행렬  
주대각선의 성분이 모두 1이며 나머지 성분은 모두 0인 정사각행렬을 단위행렬  
<img src='https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FOHtvu%2FbtqBUzNYxSz%2FhkfFHNnAv6yiMedJSqDQEk%2Fimg.jpg'>

## - 역행렬  
A와 어떤 행렬을 곱했을 때 결과가 단위행렬이라면, 이때 어떤 행렬을 A의 역행렬  
<img src='https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FdeHa6c%2FbtqBWruHorK%2Fa22CTm5PxLAl6xRMvyt9Kk%2Fimg.jpg'> 

## - 직교행렬  
행렬 A와 A의 전치 행렬을 곱했을 때 단위 행렬이 된다면, A를 직교 행렬  
<img src='https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FTSdqL%2FbtqBXF6PlGG%2FGkPYEK2KQotESao5VK4okk%2Fimg.png'>

## - 대각 행렬  
대각 성분 이외의 모든 성분이 0인 행렬  
<img src='https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FKsE28%2FbtqBWCpdJto%2FPeWwzCATz6mzkri2YFVJT0%2Fimg.png'>

## - 정방 행렬  
열과 행의 개수가 동일한 행렬을 정방 행렬이라고 합니다. n차 정방 행렬은 열과 행의 개수가 각각 n개인 행렬을 뜻합니다.  
<img src='https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FAD3cy%2FbtqBUUK89Cg%2FZUHji8FDAcewYlKFLde7zK%2Fimg.jpg'>  

## - 대칭 행렬  
정사각행렬 A와 A의 전치 행렬이 서로 같다면, A를 대칭 행렬이라고 합니다.  
<img src='https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2F1xHcH%2FbtqBUTZLeUt%2Fo9WBSkv8nPwPbwOVjrSE61%2Fimg.png'>

------

---
### LSA와 Truncated SVD
---

<img src='https://d3s0tskafalll9.cloudfront.net/media/images/GN-2-L-10.max-800x600.png'>

$$
A=USV^T
$$

LSA는 DTM이나 TF-IDF 행렬 등에 Truncated SVD를 수행합니다.  
이렇게 분해하여 얻은 행렬 3개 $(U_k,V_k^T,S)$
는 각각 '문서들과 관련된 의미들을 표현한 행렬', '단어들과 관련된 의미를 표현한 행렬' , '각 의미의 중요도를 표현한 행렬' 이라고 해석할 수 있습니다.  
위의 Truncated SVD 그림을 통해 LSA를 좀 더 자세히 이해해봅시다.  
앞의 그림과 달리 여기서는 하이퍼파라미터 t를k로 표현했다는 점을 주의합시다.

m을 문서의 수, n을 단어의 수라고 하였을 때, Truncated SVD를 통해 얻은 행렬 
$ U_k$ 는 m×k의 크기를 가집니다. 
문서의 수에 해당되는 m의 크기는 줄어들지 않았습니다.  
이때 $U_k$의 각 행은 각 문서를 표현하는 문서 벡터입니다.

반대로 $V_k^T$ 는 k×n의 크기를 가지는 행렬이 됩니다.  
단어의 수에 해당되는 n의 크기는 줄어들지 않았죠?  
이 행렬의 각 열은 각 단어를 나타내는 n 차원의 단어 벡터가 됩니다.   
A에서 단어 벡터의 크기는 m이었는데,  
$V_k^T$ 서는 k의 크기를 가지게 되었으니 벡터의 차원이 저차원으로 축소된 셈입니다.  
벡터의 차원을 저차원으로 축소하면서 잠재된 의미를 끌어내는 여러 가지 방법은 '워드 임베딩'을 배울 때 더 자세히 다룹니다.

DTM이나 TF-IDF 행렬에 Truncated SVD를 수행하고 얻은 $V_k^T$ 행렬의 k열은 전체 코퍼스로부터 얻어낸 k개의 주요 주제(topic) 라고 간주할 수도 있습니다.  

이제 실습을 통해 텍스트 데이터로부터 k개의 주제를 얻어내는 LSA 실습을 수행해봅시다.

---
# LSA와 LDA (2) LSA 실습
---

In [17]:
import pandas as pd
import numpy as np
import urllib.request
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

NLTK 데이터셋을 다운로드하지 않은 상태라면 아래의 커맨드를 통해 다운로드해 줍니다.  
NLTK는 데이터셋을 다운로드해 주지 않으면 NLTK의 도구들이 제대로 동작하지 않습니다.

In [61]:
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('stopwords')
nltk.download('omw-1.4')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


True

---
### 데이터 다운로드 및 확인
---

In [53]:
csv_filename = '/content/drive/MyDrive/aiffel/Going_Deeper/GD3/data/abcnews-date-text.csv'

urllib.request.urlretrieve("https://raw.githubusercontent.com/franciscadias/data/master/abcnews-date-text.csv", 
                           filename=csv_filename)

('/content/drive/MyDrive/aiffel/Going_Deeper/GD3/data/abcnews-date-text.csv',
 <http.client.HTTPMessage at 0x7fc6e665fcd0>)

In [54]:
data = pd.read_csv(csv_filename, error_bad_lines=False)
data.shape

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


(1082168, 2)

In [55]:
data.head()

publish_date                                      headline_text
0      20030219  aba decides against community broadcasting lic...
1      20030219     act fire witnesses must be aware of defamation
2      20030219     a g calls for infrastructure protection summit
3      20030219           air nz staff in aust strike for pay rise
4      20030219      air nz strike to affect australian travellers

publish_date는 이번 실습에 불필요하므로 headline_text만 별도로 저장합니다.

In [56]:
text = data[['headline_text']].copy()
text.head()

headline_text
0  aba decides against community broadcasting lic...
1     act fire witnesses must be aware of defamation
2     a g calls for infrastructure protection summit
3           air nz staff in aust strike for pay rise
4      air nz strike to affect australian travellers

In [57]:
text.nunique() # 중복을 제외하고 유일한 시퀀스를 가지는 샘플의 개수를 출력

headline_text    1054983
dtype: int64

In [58]:
text.drop_duplicates(inplace=True) # 중복 샘플 제거
text.reset_index(drop=True, inplace=True)
text.shape

(1054983, 1)

---
### 데이터 정제 및 정규화
---

이제 텍스트 데이터를 정제 및 정규화하는 과정을 진행해 보겠습니다.  
우선 NLTK의 토크나이저를 이용해 전체 텍스트 데이터에 대해서 단어 토큰화를 수행하고,   
NLTK가 제공하는 불용어 리스트를 사용하여 불용어를 제거합니다.

아래 과정은 다소(몇 분 정도) 시간이 걸릴 수 있으니 당황하지 마세요.

In [59]:
# NLTK 토크나이저를 이용해서 토큰화
text['headline_text'] = text.apply(lambda row: nltk.word_tokenize(row['headline_text']), axis=1)

# 불용어 제거
stop_words = stopwords.words('english')
text['headline_text'] = text['headline_text'].apply(lambda x: [word for word in x if word not in (stop_words)])

text.head()

headline_text
0   [aba, decides, community, broadcasting, licence]
1    [act, fire, witnesses, must, aware, defamation]
2     [g, calls, infrastructure, protection, summit]
3          [air, nz, staff, aust, strike, pay, rise]
4  [air, nz, strike, affect, australian, travellers]

이제 동일한 단어지만 다른 표현을 가지는 단어들을 하나의 단어로 통합(lemmatization)하는 단어 정규화 과정,  
그리고 길이가 1 ~ 2인 단어를 제거하는 전처리를 진행합니다.

In [62]:
# 단어 정규화. 3인칭 단수 표현 -> 1인칭 변환, 과거형 동사 -> 현재형 동사 등을 수행한다.
text['headline_text'] = text['headline_text'].apply(lambda x: [WordNetLemmatizer().lemmatize(word, pos='v') for word in x])

# 길이가 1 ~ 2인 단어는 제거.
text = text['headline_text'].apply(lambda x: [word for word in x if len(word) > 2])
print(text[:5])

0     [aba, decide, community, broadcast, licence]
1    [act, fire, witness, must, aware, defamation]
2       [call, infrastructure, protection, summit]
3            [air, staff, aust, strike, pay, rise]
4    [air, strike, affect, australian, travellers]
Name: headline_text, dtype: object


---
### 역토큰화 및 DTM 생성
---

DTM을 생성하는 CountVectorizer 또는 TF-IDF 행렬을 생성하는 TfidfVectorizer의 입력으로 사용하기 위해서  
토큰화 과정을 역으로 되돌리는 역토큰화(detokenization) 를 수행해보겠습니다.

In [63]:
# 역토큰화 (토큰화 작업을 역으로 수행)
detokenized_doc = []
for i in range(len(text)):
    t = ' '.join(text[i])
    detokenized_doc.append(t)

train_data = detokenized_doc

전처리 최종 결과는 train_data에 저장하였습니다. 5개의 샘플을 출력해봅시다.

In [64]:
train_data[:5]

['aba decide community broadcast licence',
 'act fire witness must aware defamation',
 'call infrastructure protection summit',
 'air staff aust strike pay rise',
 'air strike affect australian travellers']

전처리 최종 결과인 train_data는 다음에 배울 LDA 실습에서도 재사용 할 예정입니다.   
CountVectorizer를 사용하여 DTM을 생성해봅시다. 단어의 수는 5,000개로 제한하겠습니다.

In [65]:
# 상위 5000개의 단어만 사용
c_vectorizer = CountVectorizer(stop_words='english', max_features = 5000)
document_term_matrix = c_vectorizer.fit_transform(train_data)
print('=3')

=3


DTM을 생성했습니다. DTM의 크기를 확인해봅시다.  
DTM의 크기(shape)는 (문서의 수 × 단어 집합의 크기)입니다.

In [66]:
print('행렬의 크기 :',document_term_matrix.shape)

행렬의 크기 : (1054983, 5000)


---
### scikit-learn TruncatedSVD 활용
---

이제 Truncated SVD를 통해 LSA를 수행해 봅시다. 토픽의 수를 10으로 정하겠습니다.  
이는 앞서 배운 하이퍼파라미터 k에 해당되며, 행렬 $U_k^T$ 가 k × (단어의 수)의 크기를 가지도록 DTM에 TruncatedSVD를 수행합니다.

In [67]:
from sklearn.decomposition import TruncatedSVD

n_topics = 10
lsa_model = TruncatedSVD(n_components = n_topics)
lsa_model.fit_transform(document_term_matrix)

array([[ 1.20433867e-02, -3.76236339e-03,  1.82625795e-02, ...,
         2.12081426e-03, -4.97154649e-03,  1.05291312e-02],
       [ 2.90446938e-02, -1.08099218e-02,  1.82421118e-02, ...,
        -8.37770301e-03,  6.12834259e-03, -5.85453123e-03],
       [ 5.03334861e-03, -2.03395951e-03,  9.77091053e-03, ...,
        -1.90706330e-03, -4.65115165e-03,  1.77932339e-03],
       ...,
       [ 2.96525935e-02,  4.00904701e-03,  2.52927751e-02, ...,
         2.87825132e-02, -3.89642503e-03,  2.38391641e-02],
       [ 6.12798892e-02, -4.71048264e-03,  1.37696289e-01, ...,
         1.18047012e+00, -3.11826620e-01, -3.20723663e-01],
       [ 7.12309698e-02,  2.88611862e-02,  1.54453600e-03, ...,
        -4.26464475e-02,  7.82699649e-04,  9.26487073e-03]])

In [68]:
print(lsa_model.components_.shape)

(10, 5000)


행렬 $U_k^T$ 가 k × (단어의 수) 의 크기를 가지는 것을 확인할 수 있습니다.  
이제 각 행을 전체 코퍼스의 k개의 주제(topic)로 판단하고 각 주제에서 n개씩 단어를 출력해 봅시다.

In [69]:
terms = c_vectorizer.get_feature_names() # 단어 집합. 5,000개의 단어가 저장됨.

def get_topics(components, feature_names, n=5):
    for idx, topic in enumerate(components):
        print("Topic %d:" % (idx+1), [(feature_names[i], topic[i].round(5)) for i in topic.argsort()[:-n - 1:-1]])
get_topics(lsa_model.components_, terms)

Topic 1: [('police', 0.74639), ('man', 0.45353), ('charge', 0.21089), ('new', 0.14094), ('court', 0.11165)]
Topic 2: [('man', 0.69436), ('charge', 0.30096), ('court', 0.16745), ('face', 0.11375), ('murder', 0.10691)]
Topic 3: [('new', 0.83636), ('plan', 0.23684), ('say', 0.18265), ('govt', 0.11037), ('council', 0.10906)]
Topic 4: [('say', 0.73882), ('plan', 0.35413), ('govt', 0.17676), ('council', 0.13469), ('water', 0.08062)]
Topic 5: [('plan', 0.73263), ('council', 0.17978), ('govt', 0.14002), ('urge', 0.08369), ('water', 0.07095)]
Topic 6: [('govt', 0.52402), ('urge', 0.26832), ('court', 0.26699), ('fund', 0.21896), ('face', 0.17106)]
Topic 7: [('charge', 0.49449), ('court', 0.47495), ('face', 0.33189), ('murder', 0.12338), ('plan', 0.11122)]
Topic 8: [('win', 0.81395), ('council', 0.15686), ('kill', 0.12526), ('cup', 0.1127), ('australia', 0.10919)]
Topic 9: [('court', 0.6108), ('kill', 0.23054), ('accuse', 0.12639), ('face', 0.12624), ('crash', 0.11338)]
Topic 10: [('council', 0.6

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


---
#  LSA와 LDA (3) LDA
---

앞서 LSA를 통해 전체 코퍼스로부터 주요 주제인 토픽을 찾아내는 실습을 해봤습니다.  
이처럼 문서의 집합에서 토픽을 찾아내는 프로세스를 **토픽 모델링(Topic Modelling)** 이라고 합니다.  
이는 고객의 소리와 같이 많은 문서에서 주요 주제를 알아내는 일이 중요할 때 사용합니다.  

이제부터 배울 **잠재 디리클레 할당(Latent Dirichlet Allocation, LDA)** 은 토픽 모델링의 또 다른 대표적인 알고리즘입니다.  
간단하게 얘기하자면, LDA는 문서들이 토픽들의 혼합으로 구성되어 있으며, 토픽들은 확률 분포에 기반하여 단어들을 생성한다고 가정합니다.  
그리고 데이터가 주어지면, LDA는 이 가정에 따라 단어들의 분포로부터 문서가 생성되는 과정을 역추적해 문서의 토픽을 찾아냅니다.  

---
### LDA(Latent Dirichlet Allocation)의 두 가지 결과
---

위의 시뮬레이션을 통해 두 개의 행렬이 결과로서 나오는 것을 확인하셨나요? LDA는 각 토픽의 단어 분포와 각 문서의 토픽 분포를 추정해냅니다.

<img src='https://d3s0tskafalll9.cloudfront.net/media/original_images/GN-2-L-13.png'>

LDA는 각 토픽의 단어 분포, 즉 특정 토픽에 특정 단어가 나타날 확률을 추정합니다.  
예를 들어, 위의 그림의 좌측에서 'Topics'라고 적혀져 있는 부분을 보겠습니다.  
그중 초록색 토픽에는 'brain'이라는 단어가 등장할 확률이 0.04입니다.  

이번에는 그림 중앙의 'Documents'라고 적혀져 있는 부분을 보겠습니다.  
주어진 문서에는 노란색, 분홍색, 하늘색 토픽이라는 세 가지 토픽이 존재하는 것 같습니다.  
언뜻 보기에 노란색 네모가 가장 많은 것 같습니다.

이번에는 그림 우측에 'Topic proportions and assignments'를 보겠습니다.  
그 아래에 있는 막대그래프는 문서에 존재하는 토픽의 비율을 시각화한 그래프입니다.  
이 문서에는 세 가지 토픽이 존재하는데, 노란색 토픽의 비중이 가장 큰 것 같습니다.  
다시 말해 이 문서는 노란색 토픽의 단어들이 가장 많이 등장하고 있으며, 노란색 토픽일 가능성이 큽니다.  
이렇게 LDA는 단어들의 분포로부터 해당 문서의 토픽 분포를 추정하게 됩니다.



LSA는 DTM을 차원 축소하여 축소 차원에서 근접 단어들을 토픽으로 묶는 반면에,  
LDA는 단어가 특정 토픽에 존재할 확률과 문서에 특정 토픽이 존재할 확률을 결합 확률로 추정하여 토픽을 추출합니다.

---
# LSA와 LDA (4) LDA 실습
---

---
### TF-IDF 행렬 생성
---

LDA는 DTM 또는 TF-IDF를 입력으로 받을 수 있습니다. 여기서는 TF-IDF를 사용하겠습니다.  

TfidfVectorizer를 사용하여 TF-IDF 행렬을 생성해봅시다.  
단어의 수는 5,000개로 제한하겠습니다.  
TF-IDF 행렬을 생성한 후에는 행렬의 크기를 확인합시다.

In [70]:
# 상위 5,000개의 단어만 사용
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_features=5000)
tf_idf_matrix = tfidf_vectorizer.fit_transform(train_data)

# TF-IDF 행렬의 크기를 확인해봅시다.
print('행렬의 크기 :', tf_idf_matrix.shape)

행렬의 크기 : (1054983, 5000)


---
### scikit-learn LDA Model 활용
---

사이킷런의 LDA 모델을 사용하여 학습합니다.  
LSA와 마찬가지로 동일한 사이킷런 패키지이므로 앞으로 진행되는 실습 과정은 LSA와 매우 유사합니다.  
토픽의 개수는 10개로 정했습니다.  
이는 n_components의 인자값입니다. 시간이 수분 정도 소요될 것이므로 느긋하게 기다려 주세요.

In [71]:
from sklearn.decomposition import LatentDirichletAllocation

lda_model = LatentDirichletAllocation(n_components=10, learning_method='online', random_state=777, max_iter=1)
lda_model.fit_transform(tf_idf_matrix)

array([[0.0335099 , 0.0335099 , 0.0335099 , ..., 0.17024867, 0.0335099 ,
        0.0335099 ],
       [0.03365631, 0.03365631, 0.03365631, ..., 0.03365631, 0.03365631,
        0.03365631],
       [0.25184095, 0.0366096 , 0.0366096 , ..., 0.0366096 , 0.0366096 ,
        0.0366096 ],
       ...,
       [0.26687206, 0.02914502, 0.02914502, ..., 0.13007484, 0.02916018,
        0.28739608],
       [0.10378115, 0.02637829, 0.12325014, ..., 0.02637829, 0.02637829,
        0.02637829],
       [0.03376055, 0.03376055, 0.2255442 , ..., 0.03376055, 0.03376055,
        0.03376055]])

In [72]:
print(lda_model.components_.shape)

(10, 5000)


전체 코퍼스로부터 얻은 10개의 토픽과 각 토픽에서의 단어의 비중을 보겠습니다.

LSA 실습의 코드를 참고해, 아래 코드 블록에 직접 코드를 작성해 주세요.

In [73]:
# LDA의 결과 토픽과 각 단어의 비중을 출력합시다
terms = tfidf_vectorizer.get_feature_names() # 단어 집합. 5,000개의 단어가 저장됨.

def get_topics(components, feature_names, n=5):
    for idx, topic in enumerate(components):
        print("Topic %d:" % (idx+1), [(feature_names[i], topic[i].round(5)) for i in topic.argsort()[:-n-1:-1]])

get_topics(lda_model.components_, terms)

Topic 1: [('australia', 9359.06334), ('sydney', 5854.97288), ('attack', 4784.76322), ('change', 4193.63035), ('year', 3924.88997)]
Topic 2: [('government', 6344.07413), ('charge', 5947.12292), ('man', 4519.7974), ('state', 3658.16422), ('live', 3625.10473)]
Topic 3: [('australian', 7666.65651), ('say', 7561.01807), ('police', 5513.22932), ('home', 4048.38409), ('report', 3796.04446)]
Topic 4: [('melbourne', 5298.35047), ('south', 4844.59835), ('death', 4281.78433), ('china', 3214.44581), ('women', 3029.28443)]
Topic 5: [('win', 5704.0914), ('canberra', 4322.0963), ('die', 4025.63057), ('open', 3771.65243), ('warn', 3577.47151)]
Topic 6: [('court', 5246.3124), ('world', 4536.86331), ('country', 4166.34794), ('woman', 3983.97748), ('crash', 3793.50267)]
Topic 7: [('election', 5418.5038), ('adelaide', 4864.95604), ('house', 4478.6135), ('school', 3966.82676), ('2016', 3955.11155)]
Topic 8: [('trump', 8189.58575), ('new', 6625.2724), ('north', 3705.40987), ('rural', 3521.42659), ('donald',

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


---
# 텍스트 분포를 이용한 비지도 학습 토크나이저 (1) 형태소 분석기와 단어 미등록 문제
---
지금까지 띄어쓰기 단위의 단어 분포를 이용해서 단어나 문서를 벡터화하는 방법에 대해 알아봤습니다.   
이번에는 텍스트의 분포를 이용해서 토큰화를 수행하는 비지도 학습 토크나이저에 대해서 알아봅시다.

---
### 형태소 분석기의 필요성
---

한국어는 교착어입니다. 교착어란 하나의 낱말(엄밀히는 하나의 어절)이 하나의 어근(root)(혹은 어간(stem))과 각각 단일한 기능을 가지는 하나 이상의 접사(affix)의 결합으로 이루어져 있는 언어를 말합니다.   
조사라는 품사는 교착어에만 존재합니다.

이런 특성으로 인해 한국어는 영어와 달리 조사나 접사가 존재하며,  
영어처럼 띄어쓰기 단위 토큰화가 제대로 동작하지 않습니다.  

영어 토큰화를 복습해봅시다. 다음과 같이 영어 문장은 단순히 띄어쓰기를 수행해도 단어 토큰화가 제대로 동작합니다.

In [74]:
en_text = "The dog ran back to the corner near the spare bedrooms"
print(en_text.split())

['The', 'dog', 'ran', 'back', 'to', 'the', 'corner', 'near', 'the', 'spare', 'bedrooms']


그런데 한국어의 경우에는 띄어쓰기 단위로 토큰화를 하면 문제가 생깁니다.  
다음과 같은 한국어 문장을 띄어쓰기 기준으로 토큰화를 진행했다고 해봅시다.

In [75]:
kor_text = "사과의 놀라운 효능이라는 글을 봤어. 그래서 오늘 사과를 먹으려고 했는데 사과가 썩어서 슈퍼에 가서 사과랑 오렌지 사 왔어"
print(kor_text.split())

['사과의', '놀라운', '효능이라는', '글을', '봤어.', '그래서', '오늘', '사과를', '먹으려고', '했는데', '사과가', '썩어서', '슈퍼에', '가서', '사과랑', '오렌지', '사', '왔어']


위의 예제 문장에서 '사과'란 단어가 총 4번 등장했는데 모두 '의', '를', '가', '랑' 등이 붙어있어 이를 제거해주지 않으면 기계는 전부 다른 단어로 인식하게 됩니다.  
이를 해결하기 위해서 한국어는 주로 형태소 분석기를 사용합니다.  
가령, 형태소 분석기 중 하나인 Okt를 사용해서 형태소 분석을 할 경우에는 다음 결과를 얻습니다.

In [77]:
!pip install konlpy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 19.4 MB 567 kB/s 
     |████████████████████████████████| 453 kB 52.8 MB/s 


In [78]:
from konlpy.tag import Okt

tokenizer = Okt()
print(tokenizer.morphs(kor_text))

['사과', '의', '놀라운', '효능', '이라는', '글', '을', '봤어', '.', '그래서', '오늘', '사과', '를', '먹으려고', '했는데', '사과', '가', '썩어서', '슈퍼', '에', '가서', '사과', '랑', '오렌지', '사', '왔어']


---
### 단어 미등록 문제
---

하지만 기존의 형태소 분석기는 등록된 단어를 기준으로 형태소를 분류해 내어 새롭게 만들어진 단어를 인식하기 어렵다는 특징이 있습니다.

In [79]:
print(tokenizer.morphs('모두의연구소에서 자연어 처리를 공부하는 건 정말 즐거워'))

['모두', '의', '연구소', '에서', '자연어', '처리', '를', '공부', '하는', '건', '정말', '즐거워']


'모두의연구소'는 하나의 단어이지만 위의 형태소 분석 결과에서는 전부 분리된 결과를 보여줍니다.  

그렇다면 텍스트 데이터에서 특정 문자 시퀀스가 함께 자주 등장하는 빈도가 높고,  
앞뒤로 조사 또는 완전히 다른 단어가 등장하는 것을 고려해서 해당 문자 시퀀스를 형태소라고 판단하는 형태소 분석기라면 어떨까요?  

예를 들어, '모두의연구소'라는 문자열이 자주 연결되어 등장한다면 형태소라고 판단하고,  
'모두의연구소'라는 단어 앞, 뒤에 '최고', 'AI', '실력'과 같은 독립된 다른 단어들이 계속해서 등장한다면 '모두의연구소'를 형태소로 파악하는 식이지요.  

이런 아이디어를 가진 형태소 분석기가 soynlp입니다.

---
# 텍스트 분포를 이용한 비지도 학습 토크나이저 (2) soynlp
---

In [83]:
!pip install soynlp

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 416 kB 5.3 MB/s 


soynlp 깃허브에서 제공하는 예제 말뭉치를 다운로드합니다.

In [81]:
import urllib.request

txt_filename = '/content/drive/MyDrive/aiffel/Going_Deeper/GD3/data/2016-10-20.txt'

urllib.request.urlretrieve("https://raw.githubusercontent.com/lovit/soynlp/master/tutorials/2016-10-20.txt",\
                            filename=txt_filename)

('/content/drive/MyDrive/aiffel/Going_Deeper/GD3/data/2016-10-20.txt',
 <http.client.HTTPMessage at 0x7fc714e07710>)

In [84]:
from soynlp import DoublespaceLineCorpus

# 말뭉치에 대해서 다수의 문서로 분리
corpus = DoublespaceLineCorpus(txt_filename)
len(corpus)

30091

총 30,091개의 문서가 존재합니다. 공백이 아닌 문서에 한해 상위 3개의 문서만 출력해봅시다.

In [85]:
i = 0
for document in corpus:
  if len(document) > 0:
    print(document)
    i = i+1
  if i == 3:
    break

19  1990  52 1 22
오패산터널 총격전 용의자 검거 서울 연합뉴스 경찰 관계자들이 19일 오후 서울 강북구 오패산 터널 인근에서 사제 총기를 발사해 경찰을 살해한 용의자 성모씨를 검거하고 있다 성씨는 검거 당시 서바이벌 게임에서 쓰는 방탄조끼에 헬멧까지 착용한 상태였다 독자제공 영상 캡처 연합뉴스  서울 연합뉴스 김은경 기자 사제 총기로 경찰을 살해한 범인 성모 46 씨는 주도면밀했다  경찰에 따르면 성씨는 19일 오후 강북경찰서 인근 부동산 업소 밖에서 부동산업자 이모 67 씨가 나오기를 기다렸다 이씨와는 평소에도 말다툼을 자주 한 것으로 알려졌다  이씨가 나와 걷기 시작하자 성씨는 따라가면서 미리 준비해온 사제 총기를 이씨에게 발사했다 총알이 빗나가면서 이씨는 도망갔다 그 빗나간 총알은 지나가던 행인 71 씨의 배를 스쳤다  성씨는 강북서 인근 치킨집까지 이씨 뒤를 쫓으며 실랑이하다 쓰러뜨린 후 총기와 함께 가져온 망치로 이씨 머리를 때렸다  이 과정에서 오후 6시 20분께 강북구 번동 길 위에서 사람들이 싸우고 있다 총소리가 났다 는 등의 신고가 여러건 들어왔다  5분 후에 성씨의 전자발찌가 훼손됐다는 신고가 보호관찰소 시스템을 통해 들어왔다 성범죄자로 전자발찌를 차고 있던 성씨는 부엌칼로 직접 자신의 발찌를 끊었다  용의자 소지 사제총기 2정 서울 연합뉴스 임헌정 기자 서울 시내에서 폭행 용의자가 현장 조사를 벌이던 경찰관에게 사제총기를 발사해 경찰관이 숨졌다 19일 오후 6시28분 강북구 번동에서 둔기로 맞았다 는 폭행 피해 신고가 접수돼 현장에서 조사하던 강북경찰서 번동파출소 소속 김모 54 경위가 폭행 용의자 성모 45 씨가 쏜 사제총기에 맞고 쓰러진 뒤 병원에 옮겨졌으나 숨졌다 사진은 용의자가 소지한 사제총기  신고를 받고 번동파출소에서 김창호 54 경위 등 경찰들이 오후 6시 29분께 현장으로 출동했다 성씨는 그사이 부동산 앞에 놓아뒀던 가방을 챙겨 오패산 쪽으로 도망간 후였다  김 경위는 오패산 터널 입구 오른쪽의 급경사에서 성씨에

soynlp는 비지도학습 형태소 분석기이므로 기존의 형태소 분석기와는 달리 학습 과정을 거쳐야 합니다.  
전체 코퍼스로부터 응집 확률과 브랜칭 엔트로피 단어 점수표를 만드는 과정이지요.  
WordExtractor.extract()를 통해서 전체 코퍼스에 대해 단어 점수표를 계산할 수 있습니다.

In [86]:
from soynlp.word import WordExtractor

word_extractor = WordExtractor()
word_extractor.train(corpus)
word_score_table = word_extractor.extract()

training was done. used memory 4.084 Gb
all cohesion probabilities was computed. # words = 223348
all branching entropies was computed # words = 361598
all accessor variety was computed # words = 361598


---
### soynlp의 응집 확률(cohesion probability)
---

응집 확률은 내부 문자열(substring)이 얼마나 응집하여 자주 등장하는지를 판단하는 척도입니다.  
응집 확률은 문자열을 문자 단위로 분리하여 내부 문자열을 만드는 과정에서,  
왼쪽부터 순서대로 문자를 추가하면서 각 문자열이 주어졌을 때 그다음 문자가 나올 확률을 계산하여 누적 곱을 한 값입니다.  
이 값이 높을수록 전체 코퍼스에서 이 문자열 시퀀스는 하나의 단어로 등장할 가능성이 높습니다.  
수식은 아래와 같습니다.

<img src='https://d3s0tskafalll9.cloudfront.net/media/original_images/GN-2-L-14.png'>

아직은 아리송할 거예요. '반포한강공원에'라는 7의 길이를 가진 문자 시퀀스에 대해서 각 내부 문자열의 스코어를 구하는 과정은 아래와 같습니다.  

<img src='https://d3s0tskafalll9.cloudfront.net/media/images/GN-2-L-15.max-800x600.png'>

In [87]:
word_score_table["반포한"].cohesion_forward

0.08838002913645132

In [88]:
word_score_table["반포한강"].cohesion_forward

0.19841268168224552

In [89]:
word_score_table["반포한강공"].cohesion_forward

0.2972877884078849

In [90]:
word_score_table["반포한강공원"].cohesion_forward

0.37891487632839754

In [91]:
word_score_table["반포한강공원에"].cohesion_forward

0.33492963377557666

응집 확률이 최대화가 되는 내부 문자열 구간은 '반포한강공원'입니다.  
다시 말해 '반포한강공원'이 하나의 단어일 확률이 높습니다.

---
### soynlp의 브랜칭 엔트로피(branching entropy)
---

Branching Entropy는 조건부 확률의 값이 아니라 확률분포의 엔트로피값을 사용한다.  
만약 하나의 단어를 중간에서 끊으면 다음에 나올 글자는 쉽게 예측이 가능하다  
즉, 여러가지 글자 중 특정한 하나의 글자가 확률이 높다. 따라서 엔트로피값이 0에 가까운 값으로 작아진다. 하지만 하나의 단어가 완결되는 위치에는 다양한 조사나 결합어가 올 수 있으므로 여러가지 글자의 확률이 비슷하게 나오고 따라서 엔트로피값이 높아진다.

브랜칭 엔트로피(Branching Entropy) 는 확률 분포의 엔트로피값을 사용합니다.  
이는 주어진 문자열에서 다음 문자가 등장할 수 있는 가능성을 판단하는 척도입니다. 이해를 위해 퀴즈를 내보겠습니다.  
<br>
제가 어떤 단어를 생각 중인데, 한 문자씩 말해드릴 테니까 매번 다음 문자를 맞추어 보세요.  
<br>
첫 번째 문자는 '디'입니다. 다음에 등장할 문자를 맞춰보세요.  
솔직히 가늠이 잘 안 가지요? '디'로 시작하는 단어가 얼마나 많은데요. 이걸 어떻게 맞추냐구요. 정답은 '스' 입니다.
<br>
이제 '디스'까지 나왔네요. '디스' 다음 문자는 뭘까요? 벌써 정답 단어를 예측한 분도 있을 테고, 여전히 가늠이 잘 안 가시는 분도 있을 거예요.  
'디스카운트'라는 단어가 있으니까 '카'일까? 아니면 '디스코드'라는 단어가 있으니까 '코'인가?  
생각해보니 '디스코'가 정답일 수도 있겠네요. 그러면 '코'인가? '디스아너드'라는 게임이 있으니까 '아'?
전부 땡땡땡! 이 단어들을 생각하신 분들은 전부 틀렸습니다. 정답은 '플'이었습니다.
<br><br>
'디스플'까지 왔습니다. 다음 문자를 맞춰보세요. 이제 좀 명백해지는군요.  
이 정도 되면 헷갈리시는 분들은 거의 없을 거예요. 정답은 '레'입니다. '디스플레' 다음에는 어떤 문자일까요?  
너무 명백해서 문제라고 보기도 어려워졌어요. 정답은 '이'입니다.  
제가 생각한 단어는 '디스플레이'였습니다!
<br><br>

저는 지금 브랜칭 엔트로피를 시뮬레이션한 겁니다.  
브랜칭 엔트로피를 주어진 문자 시퀀스에서 다음 문자 예측을 위해 헷갈리는 정도라고 비유해 봅시다.  
브랜칭 엔트로피의 값은 하나의 완성된 단어에 가까워질수록 문맥으로 인해 정확히 예측할 수 있게 되므로 점차 줄어듭니다.  
<br>
실습해 볼게요.

In [92]:
word_score_table["디스"].right_branching_entropy

1.6371694761537934

In [93]:
word_score_table["디스플"].right_branching_entropy

-0.0

'디스' 다음에는 다양한 문자가 올 수 있으니까 1.63이라는 값을 가지는 반면,  
'디스플'이라는 문자열 다음에는 다음 문자로 '레'가 오는 것이 너무나 명백하기 때문에 0이란 값을 가집니다.  

In [94]:
word_score_table["디스플레"].right_branching_entropy

-0.0

'디스플레' 다음도 마찬가지겠죠.  
다음 문자가 '이'라는 사실이 너무나 명백해서 0이란 값을 가집니다.  
그런데 '디스플레이' 다음에는 어떤 값을 가질까요?

In [95]:
word_score_table["디스플레이"].right_branching_entropy

3.1400392861792916

갑자기 값이 급증합니다. 그 이유는 문자 시퀀스 '디스플레이'라는 문자 시퀀스 다음에는 조사나 다른 단어가 나올 수 있는 다양한 경우가 있을 수 있기 때문입니다.  
즉 하나의 단어가 끝나면 그 경계 부분부터 다시 브랜칭 엔트로피값이 증가함을 의미합니다.  
그러면 이 값으로 단어를 판단하는 것이 가능하겠죠?  
  
soynlp 패키지는 이렇게 계산한 점수를 사용해 두 가지 문자열 토큰화 방법을 제공합니다.  

---
### soynlp의 LTokenizer
---

한국어의 경우 공백(띄어쓰기)으로 분리된 하나의 문자열은 ‘L 토큰 + R 토큰; 구조인 경우가 많다.  
왼쪽에 오는 L 토큰은 체언(명사, 대명사)이나 동사, 형용사 등이고  
오른쪽에 오는 R 토큰은 조사, 동사, 형용사 등이다.  
여러가지 길이의 L 토큰의 점수를 비교하여 가장 점수가 높은 단어를 찾는 것이 L-토큰화(L-tokenizing)이다.

In [96]:
from soynlp.tokenizer import LTokenizer

scores = {word:score.cohesion_forward for word, score in word_score_table.items()}
l_tokenizer = LTokenizer(scores=scores)
l_tokenizer.tokenize("국제사회와 우리의 노력들로 범죄를 척결하자", flatten=False)

[('국제사회', '와'), ('우리', '의'), ('노력', '들로'), ('범죄', '를'), ('척결', '하자')]

---
### 최대 점수 토크나이저
---

최대 점수 토큰화(max score tokenizing)는 띄어쓰기가 되어 있지 않는 긴 문자열에서  
가능한 모든 종류의 부분문자열을 만들어서 가장 점수가 높은 것을 하나의 토큰으로 정한다.  
이 토큰을 제외하면 이 위치를 기준으로 전체 문자열이 다시 더 작은 문자열들로 나누어지는데  
이 문자열들에 대해 다시 한번 가장 점수가 높은 부분문자열을 찾는 것을 반복한다.

In [97]:
from soynlp.tokenizer import MaxScoreTokenizer

maxscore_tokenizer = MaxScoreTokenizer(scores=scores)
maxscore_tokenizer.tokenize("국제사회와우리의노력들로범죄를척결하자")

['국제사회', '와', '우리', '의', '노력', '들로', '범죄', '를', '척결', '하자']

- 띄어쓰기가 잘 되어 있는 경우: L-토큰화
- 띄어쓰기가 안되어 있는 경우: Max Score 토큰화